src/rewards/FlywheelGaugeRewards.sol

```diff
@@ -0,0 +1,173 @@
+import {SafeTransferLib, ERC20} from "solmate/utils/SafeTransferLib.sol";
+import {IFlywheelRewards} from "../interfaces/IFlywheelRewards.sol";
+
+import {ERC20Gauges} from "../token/ERC20Gauges.sol";
+
+/// @notice a contract which streams reward tokens to the FlywheelRewards module
+interface IRewardsStream {
+
+    /// @notice read and transfer reward token chunk to FlywheelRewards module
+    function getRewards() external returns(uint256);
+}
+
+/** 
+ @title Flywheel Gauge Reward Stream
+ @notice Distributes rewards from a stream based on gauge weights
+
+ The contract assumes an arbitrary stream of rewards `S` of rewardToken. It chunks the rewards into cycles of length `l`.
+
+ The allocation function for each cycle A(g, S) proportions the stream to each gauge such that SUM(A(g, S)) over all gauges <= S. 
+ NOTE it should be approximately S, but may be less due to truncation.
+
+ Rewards are accumulated every time a new rewards cycle begins, and all prior rewards are cached in the previous cycle.
+ When the Flywheel Core requests accrued rewards for a specific gauge:
+ 1. All prior rewards before this cycle are distributed
+ 2. Rewards for the current cycle are distributed proportionally to the remaining time in the cycle. 
+    If `e` is the cycle end, `t` is the min of e and current timestamp, and `p` is the prior updated time:
+    For `A` accrued rewards over the cycle, distribute `min(A * (t-p)/(e-p), A)`.
+*/ 
+contract FlywheelGaugeRewards is Auth, IFlywheelRewards {
+    using SafeTransferLib for ERC20;
+
+    // TODO add events
+
+    /// @notice the flywheel core contract
+    address public immutable flywheel;
+
+    /// @notice the reward token paid
+    ERC20 public immutable rewardToken;
+
+    /// @notice the end of the current cycle
+    uint32 public rewardsCycleEnd;
+
+    /// @notice the length of a rewards cycle
+    uint32 public immutable rewardsCycleLength;
+
+    /// @notice rewards queued from prior and current cycles
+    struct QueuedRewards {
+        uint128 priorCycleRewards;
+        uint128 currentCycleRewards;
+    }
+
+    /// @notice mapping from gauges to queued rewards
+    mapping(ERC20 => QueuedRewards) public marketQueuedRewards;
+
+    /// @notice the gauge token for determining gauge allocations of the rewards stream
+    ERC20Gauges public immutable gaugeToken;
+
+    /// @notice contract to pull reward tokens from
+    IRewardsStream public rewardsStream;
+
+    constructor(
+        ERC20 _rewardToken, 
+        address _flywheel, 
+        address _owner, 
+        Authority _authority,
+        uint32 _rewardsCycleLength,
+        ERC20Gauges _gaugeToken,
+        IRewardsStream _rewardsStream
+    ) Auth(_owner, _authority) {
+        rewardToken = _rewardToken;
+        flywheel = _flywheel;
+        rewardsCycleLength = _rewardsCycleLength;
+
+        // seed initial rewardsCycleEnd
+        rewardsCycleEnd = uint32(block.timestamp) / rewardsCycleLength * rewardsCycleLength;
```

Comment:
Not sure I follow how we calculate `rewardsCycleEnd`. Why divide by `rewardsCycleLength^2` and not `rewardsCycleLength`?

Reply:
Pretty sure the division applies first. This is basically a "round to the nearest `rewardsCycleLength` operation

Reply:
Yeah good point, BODMAS!